amp-beta-tester.php

```diff
@@ -0,0 +1,202 @@
+
+	add_filter( 'pre_set_site_transient_update_plugins', __NAMESPACE__ . '\update_amp_manifest' );
+	add_action( 'after_plugin_row_' . AMP_PLUGIN_FILE, __NAMESPACE__ . '\replace_view_version_details_link', 10, 2 );
+}
+
+/**
+ * Modifies the AMP plugin manifest to point to a new beta update if one exists.
+ *
+ * @param \stdClass $updates Object containing information on plugin updates.
+ * @return \stdClass
+ */
+function update_amp_manifest( $updates ) {
+	if ( ! isset( $updates->no_update ) ) {
+		return $updates;
+	}
+
+	$amp_version = get_plugin_data( WP_PLUGIN_DIR . '/' . AMP_PLUGIN_FILE )['Version'];
+
+	if ( isset( $updates->response[ AMP_PLUGIN_FILE ] ) ) {
+		$manifest_type = 'response';
+	} elseif ( isset( $updates->no_update[ AMP_PLUGIN_FILE ] ) ) {
+		$manifest_type = 'no_update';
+	} else {
+		return $updates;
+	}
+
+	$amp_manifest = $updates->{$manifest_type}[ AMP_PLUGIN_FILE ];
+
+	$github_releases = get_amp_github_releases();
+
+	if ( is_array( $github_releases ) ) {
+
+		foreach ( $github_releases as $release ) {
+			if ( true === $release->prerelease ) {
+				$release_version = $release->tag_name;
+
+				// If there is a new release, let's see if there is a zip available for download.
+				if ( version_compare( $release_version, $amp_version, '>' ) ) {
+					$has_zip = false;
+
+					foreach ( $release->assets as $asset ) {
+						if ( 'amp.zip' === $asset->name ) {
+							$amp_manifest->package = $asset->browser_download_url;
+							$has_zip               = true;
+							break;
+						}
+					}
+
+					$amp_manifest->new_version = $release_version;
+					$amp_manifest->url         = $release->html_url;
+
+					unset( $updates->{$manifest_type}[ AMP_PLUGIN_FILE ] );
+
+					if ( $has_zip ) {
+						$updates->response[ AMP_PLUGIN_FILE ] = $amp_manifest;
+						break;
+					}
+				}
+			}
+		}
+	}
+
+	return $updates;
+}
+
+/**
+ * Fetch AMP releases from GitHub.
+ *
+ * @return array|null
+ */
+function get_amp_github_releases() {
+	$raw_response = wp_remote_get( 'https://api.github.com/repos/ampproject/amp-wp/releases' );
+	if ( is_wp_error( $raw_response ) ) {
+		return null;
+	}
+	return json_decode( $raw_response['body'] );
+}
+
+/**
+ * Replace the 'View version details' link with the link to the release on GitHub.
+ *
+ * @param string $file Plugin file.
+ * @param array  $plugin_data Plugin data.
+ */
+function replace_view_version_details_link( $file, $plugin_data ) {
+	$plugin_version = $plugin_data['Version'];
+
+	if ( is_pre_release( $plugin_version ) ) {
+		?>
+		<script>
+			document.addEventListener('DOMContentLoaded', function() {
+				const links = document.querySelectorAll("[data-slug='amp'] a.thickbox.open-plugin-details-modal");
+
+				links.forEach( (link) => {
+					link.className = 'overridden'; // Override class so that onclick listeners are disabled.
+					link.target = '_blank';
+					<?php
+					if ( isset( $plugin_data['url'] ) ) {
+						echo "link.href = '" . esc_url( $plugin_data['url'] ) . "';";
```

Comment:
The `esc_url()` function intends an HTML target, rather than a JS one. This should be better:

```suggestion
						printf( 'link.href = %s;', wp_json_encode( esc_url_raw( $plugin_data['url'] ) ) );
```